### ISSUE TITLE
Dependency management needed?

### ISSUE THREAD

**Issue by joostfarla** [jaws-framework/JAWS#167]
Great to see new modules popping up every day. This is really exciting!

For the JAWS ecosystem to be able to grow and stabilize, we need to think about how to handle dependencies. The current module implementation is very basic and does not take versioning and underlying dependencies into account (see: https://github.com/jaws-framework/JAWS/issues/148).

Things we have to start debating about:

* Do we need to store dependency configuration in the project (jaws.json) and module (awsm.json) configurations? I think we do, because:
  * ideally you want to keep vendor packages out of version control. We could implement a script which walks through the dependency graph and installs any missing dependencies and upgrades versions where needed/possible. This is crucial when working in development teams.
  * otherwise there’s no place where you can define underlying dependencies including compatible versions or version ranges.

* How to handle underlying AWSM dependencies, i.e. module X requires module Y, which requires module Z? This would definitely stimulate creating small components focused on one specific task. I would suggest flattening the dependency graph (i.e. store all modules on the same directory level) to keep things simple for other tasks like deployments. Taking the recursive approach (like NPM does) would make things too complex. Downside (or is this maybe an upside?) of this approach is that you cannot have multiple versions of the same AWSM module in your project which could cause dependency conflicts.

* Is the current project structure suitable for this? I would suggest splitting vendor packages from your application code/modules. Otherwise it’s too hard to do verification on which modules should be (un-)installed. Moreover I find it more clean to separate these.

* Do we need to support pre/post install-scripts for any custom checks or build steps a module needs. At this moment the installer checks for the presence of a package.json file and runs npm install. This could be made more generic (especially for non-node modules in the future).

I’m having tons of other considerations, but I guess we have to agree on the basics first.

Looking forward to your thoughts! I would be happy to start implementing this once we agree on the implementation.

**Comment by jwulf** [comment]
We're going to replicate npm. I reckon we should just leverage it.

**Comment by dekz** [comment]
+1 this to me seems like an absolute must. 

> Do we need to support pre/post install-scripts for any custom checks  

There is a definite need for user input to configure an AWSM module. DynamoDB table names to name one. The likely place for this is on `module install/update` or `deploy`. 

> We're going to replicate npm  

I don't think we should be quick to accept NPM, we should consider the fact that Lambdas support multiple languages and what the projects Goals towards this is.

**Comment by joostfarla** [comment]
> We're going to replicate npm. I reckon we should just leverage it.

I agree we must not reinvent the wheel, we definitely need to leverage existing modules wherever we can. However we first have to agree on the way we want things to work before considering using NPM. Like @dekz said, it's also important things will be made compatible with other platforms (not saying this is not possible with NPM).

**Comment by ac360** [comment]
This is one we've been thinking about a lot.  Some important considerations:

* aws-modules are (mostly) lambdas and lambdas are your back-end code.
* aws-modules will be edited, unlike npm modules.  Many will consider them templates

Therefore, I believe that aws-modules should be checked into version control.  They are your app's code, after all.

Also, given the above considerations, splitting them might not be best either.  If an aws-module contains a lambda, or an API Gateway endpoint, it will likely be edited.

```
back
   aws_modules
      app
      vendor
```
The above points are why we kept it simple with the below structure because aws-modules will be edited:
```
back
   aws_modules
      module1
```

I am also a huge fan of pre/post-install scripts.  What I have been thinking about a lot is adding some type of callback/event emitter within JAWS.  This would make it more extensible.

**Comment by joostfarla** [comment]
I believe it's dangerous to mix up the concepts of modularity and templating. These are in essence two different concepts. When I think of a module, I think of something that is re-usable, loosely coupled and highly cohesive. Templates, on the other hand, are just boilerplate code but could be a valuable (probably the most important) part of AWSM modules though.

If AWSM modules are meant to be edited, what is the purpose of versioning and the `module update` command anyway? I'm wondering whether this is so different than other modularized application frameworks. Modules always have to be written in a way that allows developers to pass customization options (or use environment variables) for configuration and extensibility.

Take for example the `jaws-core-js` module. While it evolves (adding middleware layer and stuff like that), it will more and more become a crucial part for projects. When fixing bugs, you might want people to profit from these fixes as fast as possible (i.e. without having to update manually). When you allow people to edit this module, an update would cause conflicts or overwrite the customizations, which will end up in unmaintainable code. Of course it's important to think about the risks (e.g. backwards compatibility) but that can be solved by following a clear and robust versioning strategy.

**Comment by ac360** [comment]
@joostfarla I agree with everything you've said.  

I'm very interested in ideas about how to write modules in ways that allow developers to pass customization options for configuration and extensibility.  But I don't know if there is a silver bullet solution here.  I think hackers like to hack.  Especially if the modules themselves handle the application's core logic.  My knee-jerk reaction is that I would immediately start modifying the module.

Versioning and module update don't appear to have a strong purpose.  Module update is provided for convenience to delete and replace an aws-module from its repo, but also creates confusion.

Also, we still have npm, which can and should handle node.js code that is not meant to be modified.  The `jaws-core-js` module problem could be handled by simple making it an npm module, instead of an aws-module, and define it in `back/package.json`.  This way, other lambda-supported-languages could have their own `jaws-core` modules written as a module supported by those languages' best package manager.

**Comment by joostfarla** [comment]
Moving `jaws-core-js` to a NPM package sounds like a great idea!

I think this approach could apply for more use cases. In my case, currently I'm working on an OAuth implementation in JAWS. The authentication logic is something that applies on all my API gateway endpoints, so ideally this would be middleware and separate from all application modules. I don't want people to edit the OAuth logic, but I do want to expose some options to provide project-specific configuration.

In this use case I could create a NPM package, add it in `/back/package.json`, load the middleware in the application code and pass it a configuration object. The question that remains is how do I get the OAuth endpoints in place for API gateway / Lambda. This could be done by providing a second (AWSM) module package which can be loaded by `module install`, but we have to think about the confusing this might cause. Another option could be to provide some generator functions which automates the generation and merging of application AWSM templates from boilerplate templates, which reside in the NPM package.

Another thing I'm not very fond of is the way multiple package.json files are spread across the project. This makes installing or updating the whole project quite painful. I would suggest working from a single package.json file in the root of the project. With changing the way dependencies are composed, we might be able to solve this also?

**Comment by ac360** [comment]
@joostfarla:  We thought hard about enforcing only 1 package.json in the `back/` root.  However, this introduces other complexities...

* The biggest is this makes it more complex to publish aws-modules.  If `package.json` is located within the aws-module itself, then it is easier to publish because you can do so from within the project you are working on.  But, as soon as you force `package.json` to be located outside of the aws-module within a JAWS project, then you have to have a separate codebase where you maintain your module, with `package.json` located inside.  We thought this would be a huge disincentive for people creating aws-modules.  

* Second, now we have to merge the aws-module's package.json with the master package.json upon install, and prompt the user for conflicts/multiple versions.  We weren't sure we wanted to get into that...

So, we opted to allow multiple package.json locations, in `back` or in aws_modules themselves.

As of now, you have 2 options:

* Make an npm package separately that you install into `back/package.json`.  This is definitely the place to install npm modules you want to use across all your aws-modules/lambdas.
* Make your aws-module look like this:
```
aws_modules
   oauth
      lib
         shared_code.js
      lambda1
      lambda2
```
You can require in your shared_code.js into other modules using relative paths.

**Comment by joostfarla** [comment]
Thanks for pointing this out! However, it still feels not maintainable enough.

Some idea that just crossed my mind. What do you think of the following project structure (for JS)?

```
/back
     /aws_modules
          /oauth2 => copy from or symbolic link to /node_modules/jaws-oauth2/awsm
/node_modules
     /jaws-oauth2/
          /awsm
               awsm.json
               /oauth2
                    /token
                         awsm.json
                         handler.js
                         index.js
          /lib
              /shared-code
          /node_modules
             /dependencies
          package.json
```

Now the workflow is clear. First, we can import the package using NPM (from the root)
```npm install jaws-oauth```

We could provide a CLI command to publish the aws_module in the back directory:
```jaws publish jaws-oauth```

This could copy over the files in the `awsm` directory (after which it can be customized), but for easy development/publishing it can also be used with a symlink parameter:
```jaws publish jaws-oauth --symlink```

For publishing AWMS modules multiple times (e.g. for CRUD module or something), a custom module name could be provided:
```
jaws publish jaws-crud users
jaws publish jaws-crud orders
```

This CLI command could possibly be run on post-install (when it applies).

With this structure I think we could achieve a lot:
* We have only one package.json in our project
* Vendor packages can be ignored by version control
* When using the symlink option, you are able to publish the package directly.
* We can use all the goodness of NPM, like versioning and dependency resolving
* There is no merging of package.json files, because there's only one file in the root. 
* This also could solve the problem with search paths (see https://github.com/jaws-framework/JAWS/issues/168)?

I'm curious about your thoughts!

**Comment by ac360** [comment]
Hmm, it is an interesting approach because it is backwards to our previous thinking...

* We can't use symlink.  I got swamped with issues about this in JAWS V0.  The biggest issue is that your global namespace gets crowded/polluted.  As soon as you have more than 1 JAWS project, your symlinks become a total mess.

* We'd have to implement a similar backwards approach for the other programming languages via their respective package managers.  Not sure how this would work yet.

I'm going to keep thinking about this.  Maybe there is something here.  Nice thinking @joostfarla

**Comment by ac360** [comment]
I think this is an intriguing way to solve our template + module dilemma.  I'm still thinking it through and dumping more thoughts in here...

I guess the module author could deal with the symlinking if they wanted to.  But everyone using the module should not have to do this.  It's unsustainable.

What do you think of this @joostfarla:

I'm favoring putting awsm.json at the root level, ATM...

**Module Structure**
```
     /awsm-oauth2
          /awsm
               /oauth2
                    /token
                         awsm.json
                         handler.js
                         index.js
          /lib
              /shared-code
          /node_modules
             /dependencies
          package.json
          awsm.json
```

**Project Structure**
```
/jaws
    /back
        /aws_modules
        /node_modules
        package.json
```

**Module Creation Flow:**

* `jaws module create awsm-users list`
* Creates  module in aws_modules
* Prompts the User asking them if they want to create/update a respective npm module for this aws_module.
* If User says ok, creates awsm-users module in node_modules

**Module Publishing Flow:**

* Perhaps we could create a command to persist what is in aws_modules back to the node_module/awsm folder.  However, it's nice that the author of a module has effectively 2 different aws_modules locations for 1 module.  They will want to put their module to use and customize their module in aws_modules/module/.  While for publishing to the community, they can have simple, starter code in their `node_modules/module/awsm.


**Module Installation Flow:**

* `npm install awsm-users`
* Copy node_modules/awsm-users/awsm into aws_modules/awsm-users

**Comment by doapp-ryanp** [comment]
I haven't had time to read this super throughly, however in some initial projects I've made with jaws I found putting a package.json at the root level of the `back` dir has worked well for node mods that are shared across resource dirs.  This mechanism also lends its self really well to js packaging (via browserify) because of how [node looks for modules](https://nodejs.org/api/modules.html#modules_loading_from_node_modules_folders).

**Comment by jwulf** [comment]
On the "let's not use npm because other languages" - other languages also have dependency management tools (like ruby gems, gradle for java, etc) and they can be leveraged for those languages when they show up. 

We could have awsms as packages that install globally and follow a convention to expose scaffolding.  Atm jaws (like Express) scaffolds a new project from a globally installed module. What about doing "jaws install <some awsm>" and having:

0. jaws check if that module is already globally installed
1. If not, npm install it 
2. If it is, scaffold it into the current project

And have a jaws command to scaffold a new awsm with the required layout to expose it as "scaffoldable".

**Comment by jwulf** [comment]
That was meant to be:

jaws install _some module_

**Comment by ac360** [comment]
@jwulf :+1: 

`jaws install <module_name>` is going to need a flag to tell it what language/package manager to use then.  For example, `jaws install <module_name> --js` will install from npm.  Perhaps the default can be npm.

If `jaws install` wraps `npm install`, we can then perform pre/post operations without forcing the module author to put pre/postinstall scripts in their package.json

**Comment by jwulf** [comment]
Making the default npm is good. Sane defaults plus extensibility. Build to the immediate use case. Release early, release often.... all that good stuff...

**Comment by joostfarla** [comment]
@ac360 @jwulf @doapp-ryanp thanks for your thoughts!

**Module Structure**

I agree it's better to put `awsm.json` in the root of the module package instead of the `awsm` dir!

**Project Structure**

It's conventional for all major frameworks to put the (dependency) configuration file in the root of your project. I think it will cause a lot of confusion when developers have to switch to the `/back` dir en run `npm install` there. I don't see any disadvantages of using one shared `package.json` for both back- and frontend code. 

Since we're reconsidering the current project structure, I would als like to propose to get rid of the `back` and `front` directories. So, like this (where `jaws` is the project directory):

```
/jaws
        /aws_modules
        /node_modules
        package.json
```

These directories have no purpose at this time. If JAWS is adding support for frontend integration in the future, we can always add something like a `public` directory.

**Module Creation Flow**

> *Offer a --npm option it they want to create/update a respective npm module for this aws_module.
> *If User says ok, creates awsm-users module in node_modules

I'm not sure if one would like to have a directory auto-created in `node_modules`. The `node_modules` directory is meant to leave untouched. When someone is creating a new NPM package, I guess they better might follow the common NPM workflow (which is creating a directory outside of the project and use `npm link` to make it available during development).

We could create a set of commands to simplify synchronizing the project module with the module in the vendor package. These could be the same commands as the ones used when installing a vendor package of some 3rd party.

**Module Installation Flow**

> jaws install awsm-users This wraps npm install and performs an npm installation.

I don't see why we would wrap NPM instead of letting developers use NPM directly? This would mean JAWS needs support for all new languages coming in the future. NPM (and most other package managers) already support `postinstall` scripts, so this might seem unnecessary. The only thing JAWS has to do is expose CLI commands for importing the awsm dir and merging in the CF code. Of course we need to document how it should be used with `postinstall` configuration. The advantage of this approach is that each module can have different configuration on which commands need to be run after installation.

I really think we're getting somewhere! :D

**Comment by ac360** [comment]
@joostfarla @doapp-ryanp @jwulf  What do you think of these commands:

* **jaws create** - Creates a module.  Defaults to nodejs runtime.  Doesn't make anything in node_modules.
* **jaws sync** - Persists aws_module to node_modules/module/awsm folder.  Creates a node_module if it doesn't exist.  An `-a` option allows you to persist only awsm.json.  Defaults to nodejs runtime. 

We can make a special command for npm postinstall (and other languages).  Relying on npm install might be easiest.  Though, developers will have to remember to add it.

I'm going to await some more feedback on the directory structure too.  Perhaps it is time to nix the front folder.

**Comment by ac360** [comment]
Alright, we're going to try this in v1.3.

We're going to refactor the aws-module specs per the above conversation.

We are also going to get rid of the front/back folders and go with the project structure @joostfarla recommends above. 

All of this should result in a much cleaner architecture, better modules, better cross-module requiring, and better workflow.  We'd love some help on this!  We've got to finish this and smooth out the bugs before re:invent, which happens next week.  

We're going to do the project structure refactor first in branch v1.3, then we will branch off from that to do the aws-module refactor in a v1.3-awsm branch.

**Comment by joostfarla** [comment]
Great to see this is going to be implemented on this short notice!

About the commands, ik think we have to elaborate a bit more on the tasks that could be needed on `postinstall`. I'm not sure if one simple command will be enough. Some examples:

* As mentioned before, some awsm modules might be candidate for importing into the project multiple times. If the module contains a boilerplate for CRUD for example, this could be imported multiple times with multiple names:

```
jaws import awsm-crud users
jaws import awsm-crud orders
 ```

* Some modules need user input on import, like DynamoDB table names (as @dekz said) or path prefixes. There could be some kind of interactive installer process for this.

* Running preprocessors or other build steps?

Any more ideas on tasks which could run on `postinstall`?

**Comment by dekz** [comment]
Is there a possible use case that I would want to mix languages in my Lambda project?

> awsm-users, node
> awsm-geo, java

What does this mean for the root `package.json`?

**Comment by ac360** [comment]
@dekz Some languages will be more efficient at certain tasks than others.  Package.json will sit alongside other package manager files for the other languages.

@joostfarla:  I'd like to get this refactor completed before re:invent.  Unless we can come up with a super elegant way to collect user input on install and deal with path prefixing, I'd like to push that stuff a bit further down the road (v1.4).  My main concern is getting the minimum requirements working well and making sure the aws-module format is as good as it can be (and that is already overwhelming me).  Could we focus on the bare minimum for now, and build everything so that we can add more postinstall tasks later?

How about this for the initial postinstall tasks:

* Copies npm_module/awsm/ to the aws_modules folder
* Adds module's CloudFormation template into resources-cf.json

**Comment by binoculars** [comment]
Wow lots of good discussion here. I've been confused for a while by the module/template naming convention. Looking at it now, it makes a lot more sense to call them templates rather than modules. However, as a user I would prefer modules--mainly because of reasons previously stated like updates.

Are there plans for some sort of super package manager? I.e. a site (built with JAWS of course) that would mostly serve as a pointer system to other package managers (npm, pip, etc)? The jaws cli could be used to register these. Sounds like the direction it's going--can anyone clarify?

**Comment by ac360** [comment]
@binoculars Yes!  That's what we're thinking.  It may not be a full package manager, but it will be a registry/pointer system.  Using it, you can easily find an aws-module in your flavor.  

`jaws search awsm-users`

Found: 3 results
- awsm-users (javascript)
- awsm-users (java)
- awsm-users (python)

**Comment by joostfarla** [comment]
@ac360 You're right, we have to get the minimum set of features released asap! Today is a busy day for me for work, but I'll see where I can help the rest of the week! Is it an idea to split the changes in several small issues so that we know who's working on what and where help is needed?

About the `postinstall` stuff, I agree these things don't have high priority. The only thing for now is trying to achieve agreement on how we can customize the postinstall proces and extend it with new tasks in the future. In a way we do not have to change the configuring and CLI syntax in a backwards-incompatible manner down the road.

> How about this for the initial postinstall tasks:
> * Copies npm_module/awsm/ to the aws_modules folder
> * Adds module's CloudFormation template into resources-cf.json

What I'm mainly wondering: are these 2 separate commands or is this one generic command which reads configuration from the awsm.json file? We could provide a set of settings in awsm.json to configure the `postinstall` requirements for that package and extend it in the future with settings for things like interactive installing.

**Comment by doapp-ryanp** [comment]
@joostfarla, for nodejs awsm modules, I envision them implementing the npm `postinstall` and calling just 1 command: `jaws module postinstall`.  This would also apply to other pkg managers like pip etc.

`jaws module postinstall` then:
*  Copies npm_module/awsm/ to the aws_modules folder (for nodejs)
*  Adds module's CloudFormation template extension points (`resources.cloudFormation.*) into resources-cf.json

To solve the user input scenario, I propose we leverage `awsm.json:lambda.envVars` that exists in EACH lambda action folder.
*  jaws iterates through each module subdir that has a `awsm.json`.  If it has non empty `lambda.envVars` it goes thru each. If key is NOT a pre-defined `JAWS_*` env var it PROMPTS the user to set the value right then.  Beauty of this is it uses the `jaws env set` command under the covers.  I propose we only do this for one stage+region (maybe we prompt them for this).  ALTERNATIVELY if we detect `lambda.envVars` the output of `jaws module postinstall` could simply be telling the end user "the following env vars need to be set: BLAH, FOO, BAR".  I was thinking we could change the `lambda.envVars` list to be {key:"SOME_ENV",desc:"some description"} HOWEVER I think we should really force awsm module creators to make good README.md's.

These steps above solve the following:

*  Module creators only have to specify on call in their package managers post install hook (npm,pip,etc)
*  We have lots of extensibility if  `jaws module postinstall` is looking at the modules `awsm.json` files to figure out what todo. This includes, but not limited to, knowing what jaws version the awsm module requires (this is a to be added attr in awsm.json IMO)
*  To @dekz point, this will allow a project to have a mix of languages.  This however will NOT allow one awsm module to have a mix of languages.  I think this is a reasonable assertion and will make jaws usable/maintainable going forward.  One specific example this allows for is...
*  The to be created awsm registry @ac360 spoke of can be totally language agnostic. It will just be a registry of metadata about awsm modules.  It will just know names,keywords,desc,source control URL and runtime.

Anyone see issue with this? Especially is there any use case hole in my user input collection scenario?

**Comment by doapp-ryanp** [comment]
I did think of something, we may need to collect input to populate CloudFormation values in the awsm CF extension points.

**Comment by shortjared** [comment]
`If key is NOT a pre-defined JAWS_* env var it PROMPTS the user to set the value right then.`

Something to think about on this piece, because you can ENV vars across so many different stages and regions, where does this get set? Do you push them through stage / region selectors? Default to local env? 

I think I am more in favor of "Module requires the following {ENV_VARS} to be set." messaging. Another option is the lambda.envVars list is something like `{key:"SOME_ENV",desc:"some description", required:true}` a true benefit that you could check this at deploy time and see if the env file contains everything needed for a successful deploy to prevent botched deploys and transitions from stage to stage. I have certainly done an ENV var in dev, then forgot to put it into staging, or worse, production.

**Comment by remicastaing** [comment]
Would it be possible to handle the `front` part as a simple (and single) aws module. It could be hosted inside of awsm_module.  In his `awsm.json` file, you would find something like `s3: {...}` instead of `lambda:{...}` and It would embed its own builder (gulp/grunt script).
The front aws module would be unique because, unlike lambda function you couldn't compose the front end with aws module as component of the shelf.

**Comment by dekz** [comment]
As long as the awsm registry is written as a JAWS project, I have no issues ;).

I have a use case for the CLoudFormation resource to be be effected by user input and that is a database table name. `awsm-users` assumes `jaws-users`, I guess it could create `myproject-users` if that variable is available at CloudFormation time and ENV runtime. 

Some use cases might include Bring Your Own Database resources, eg.
```
> Allocate a DynamoDB?
> yes [no]
> Please enter Preallocated DynamoDB Table Name:
> not-jaws-users
... not-jaws-users is input into IAM role but not added to resources to deploy ...
```

Or is this use case too extreme?

**Comment by joostfarla** [comment]
@ac360 has been working on an unforeseen problem with the new dependency approach. Just putting some thoughts here (which I've already expressed in the Gitter chat).

The problem here is that the module “postinstall" command duplicates the module’s awsm directory to your project’s aws_modules dir. Both locations have access to different dependencies due to the recursive dependency approach of NPM. This results in not being able to access the module's underlying dependencies from the aws_modules directory.

What I’m thinking atm is that we have to try to work towards an architecture where there is no need to access the module’s underlying dependencies from the aws_modules path. This could be done by splitting handler.js and index.js, where handler.js resides in the project's aws_modules dir and index.js resides in the module's path (within /node_modules). This would require a small change in the AWSM module structure.

An extra advantage is that this file also becomes versioned and maintained while the module evolves. A drawback could be that the module's logic cannot be customized. However, index.js could pass customization options to index.js, which makes module owners more aware of the degree of flexiblity they have to implement.

I think we have to stay away from customizing the way NPM handles dependencies by nature. This can cause a serious lot of headaches in the future when implementing stuff in the future. NPM v3 is supporting a flattened dependency structure, but this is all very new and probably an unwanted requirement for now.

**Comment by jwulf** [comment]
What about this?

```
npm install -g awsm-middleware
npm install -g awsm-loggly

cd ~/workspace/jaws-project/
jaws import middleware
jaws import loggly
```

`jaws import` scaffolds the lambda in the current project, and also invokes `npm install` inside the scaffolded lambda.

**Comment by ac360** [comment]
Fixed in v1.3 394cf7e  Thanks everyone and @joostfarla
